compiler/scoqc.ml

```diff
@@ -0,0 +1,124 @@
+    | _ -> ()
+
+let parse ~st pa =
+  let mode = Option.map (fun _ -> Vernacinterp.get_default_proof_mode ()) st.Vernacstate.lemmas in
+  Vernacstate.(Parser.parse st.parsing Pvernac.(main_entry mode)) pa
+
+let execute = Vernacinterp.interp
+
+let rec cloop ~st pa =
+  match parse ~st pa with
+  | None ->
+    st
+  | Some stm ->
+    let st = execute ~st stm in
+    (cloop [@ocaml.tailcall]) ~st pa
+
+let touch_file f =
+  let out = open_out f in
+  output_char out ' ';
+  close_out out
+
+let output_native_objects =
+  match Coq_config.native_compiler with
+  | NativeOn _ -> true
+  | NativeOff -> false
+
+let save_library ldir in_file ~out_file =
+  let out_vo = Option.default (Filename.(remove_extension in_file) ^ ".vo") out_file in
+  let todo_proofs = Library.ProofsTodoNone in
+  Library.save_library_to todo_proofs ~output_native_objects ldir out_vo;
+
+  let out_aux =
+    let d, f = Filename.dirname in_file, Filename.basename in_file in
+    let f = Filename.(remove_extension f) in
+    d ^ Filename.dir_sep ^ "." ^ f ^ ".aux"
+  in
+  touch_file out_aux
+
+let start_glob ~in_file =
+  let out_vo = Filename.(remove_extension in_file) ^ ".vo" in
+  Dumpglob.push_output Dumpglob.MultFiles;
+  Dumpglob.start_dump_glob ~vfile:in_file ~vofile:out_vo;
+  (* Dumpglob.dump_string ("F" ^ Names.DirPath.to_string ldir ^ "\n"); *)
+  ()
+
+let touch_vos ~in_file =
+  let out_vos = Filename.(remove_extension in_file) ^ ".vos" in
+  Stdlib.open_out out_vos |> Stdlib.close_out
+
+let compile ~args ~injections ~extra =
+  let in_file = match extra with
+    | [] -> exit 0
+    | x :: _ -> x
+  in
+  let f_in = open_in in_file in
+  let () = start_glob ~in_file in
+  let libname = dirpath_of_file in_file in
+  let () = Coqinit.start_library ~top:libname injections in
+  let st = Vernacstate.freeze_interp_state ~marshallable:false in
+  let pa = Pcoq.Parsable.make (Gramlib.Stream.of_channel f_in) in
+  let _st = cloop ~st pa in
+  (* Compact the heap, just in case. *)
+  Gc.compact ();
+  (* print_st_stats st; *)
+  let () = save_library libname in_file ~out_file:None in
+  let () = touch_vos ~in_file in
+  Dumpglob.end_dump_glob ();
+  ()
+
+let rec ignore_extra args = match args with
+  | "-async-proofs-tac-j" :: _ :: rem ->
+    ignore_extra args
+  | [] -> []
+  | r :: rem ->
+    r :: (ignore_extra rem)
+
+let () =
+  Flags.quiet := true;
+  System.trust_file_cache := true;
+  Coqinit.init_ocaml ();
+  let usage = { Boot.Usage.executable_name = "coqc"
+              ; extra_args = ""
+              ; extra_options = ""
+              } in
+  try
+    let parse_extra l = (l, []) in
+    let args, extra = Coqinit.parse_arguments ~parse_extra ~usage () in
+    let extra = ignore_extra extra in
+    ignore (Feedback.add_feeder fb_handler);
+    let injections = Coqinit.init_runtime args in
+    compile ~args ~injections ~extra
+  with exn ->
+    Format.eprintf "Error: @[%a@]@\n%!" Pp.pp_with (CErrors.print exn)
```

Comment:
Yes, that's a bug hehe; bear with me, I was not planning to submit this, but at the WG there was some discussion about how Coq looks without the STM, and in fact, I had this branch since years so I thought it'd be a good idea to share it.

Let me amend.

Reply:
Feel absolutely free to push / play with this.